### PR TITLE
Telldus runtime detection

### PR DESCRIPTION
**Pull request by Risca**
This change will allow POSIX systems to build without telldus-core support available on the build server, but still be able to detect telldus-core in runtime.
This change also make Tellstick.{h,cpp} to not be skipped during normal compilation and cause possible regressions (see e.g. #2715)
Lastly, this change also removes the checked in copy of telldus-core that seems to have been there since the dawn of time.

### CODE REVIEW

File: hardware/TelldusFunctions.h [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+
+	int (WINAPI *TurnOn)(int intDeviceId);
+	int (WINAPI *TurnOff)(int intDeviceId);
+	int (WINAPI *Dim)(int intDeviceId, unsigned char level);
+	int (WINAPI *Methods)(int id, int methodsSupported);
+
+	int (WINAPI *GetNumberOfDevices)();
+	int (WINAPI *GetDeviceId)(int intDeviceIndex);
+
+	char * (WINAPI *GetName)(int intDeviceId);
```
Comment (Risca) [inline hardware/TelldusFunctions.h:35]: I only added the functions used by `Tellstick.cpp`

File: hardware/TellstickFactory.cpp [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+        }
+    }
+
+    if (g_symbolsLoaded) {
+        CTellstick* obj = new CTellstick(g_functions, ID, repeats, repeatInterval);
+        if (obj) {
+            *tellstick = obj;
+            return true;
+        }
+    }
```
Comment (Risca) [inline hardware/TellstickFactory.cpp:103]: At first, I loaded the symbols again on every call to `Create()`, but `boost::signals2` was giving be `SIGABRT`s when I did that so I added the "caching" feature.

Comment (Risca) [reply hardware/TellstickFactory.cpp:103]: Old crash log from before "caching" change:
```
$ ./domoticz 
2018-09-24 21:42:06.668  Status: Domoticz V4.10008 (c)2012-2018 GizMoCuz
2018-09-24 21:42:06.668  Status: Build Hash: d0cbc396-modified, Date: 2018-09-23 21:50:28
2018-09-24 21:42:06.683  Status: Startup Path: /home/risca/Documents/programming/domoticz/
2018-09-24 21:42:06.707  Status: PluginSystem: Started, Python version '3.7.0'.
2018-09-24 21:42:06.707  Tellstick: dynamically loaded telldus-core
2018-09-24 21:42:06.709  Active notification Subsystems: gcm, http (2/13)
2018-09-24 21:42:06.710  Status: WebServer(HTTP) started on address: :: with port 8080
2018-09-24 21:42:06.714  Error: WebServer(SSL) startup failed on address 0.0.0.0 with port: 443: bind: Permission denied
2018-09-24 21:42:06.714  Error: WebServer(SSL) check privileges for opening ports below 1024
2018-09-24 21:42:06.714  Status: Proxymanager started.
2018-09-24 21:42:06.714  Starting shared server on: :::6144
2018-09-24 21:42:06.715  Status: TCPServer: shared server started...
2018-09-24 21:42:06.715  Status: RxQueue: queue worker started...
domoticz: /usr/include/boost/signals2/detail/lwm_pthreads.hpp:60: void boost::signals2::mutex::lock(): Assertion `pthread_mutex_lock(&m_) == 0' failed.
2018-09-24 21:42:18.219  Error: Domoticz(pid:21031, tid:21041('MainWorker')) received fatal signal 6 (Aborted)
2018-09-24 21:42:18.219  Error: siginfo address=0x3e800005227, address=0x7fd3b65edd7f
2018-09-24 21:42:18.258  Tellstick: dynamically loaded telldus-core
2018-09-24 21:42:26.237  Error: Thread 10 (Thread 0x7fd3ab7fe700 (LWP 21041)):
2018-09-24 21:42:26.237  Error: #0  0x00007fd3b6930d42 in waitpid () from /usr/lib/libpthread.so.0
2018-09-24 21:42:26.237  Error: #1  0x0000558e8d80fb22 in dumpstack_gdb (printAllThreads=false) at /home/risca/Documents/programming/domoticz/main/SignalHandler.cpp:260
2018-09-24 21:42:26.237  Error: #2  0x0000558e8d8100b9 in dumpstack (info=0x7fd3ab7fd030, ucontext=0x7fd3ab7fcf00) at /home/risca/Documents/programming/domoticz/main/SignalHandler.cpp:379
2018-09-24 21:42:26.237  Error: #3  0x0000558e8d81032c in signal_handler (sig_num=6, info=0x7fd3ab7fd030, ucontext=0x7fd3ab7fcf00) at /home/risca/Documents/programming/domoticz/main/SignalHandler.cpp:462
2018-09-24 21:42:26.237  Error: #4  <signal handler called>
2018-09-24 21:42:26.237  Error: #5  0x00007fd3b65edd7f in raise () from /usr/lib/libc.so.6
2018-09-24 21:42:26.237  Error: #6  0x00007fd3b65d8672 in abort () from /usr/lib/libc.so.6
2018-09-24 21:42:26.237  Error: #7  0x00007fd3b65d8548 in __assert_fail_base.cold.0 () from /usr/lib/libc.so.6
2018-09-24 21:42:26.237  Error: #8  0x00007fd3b65e6396 in __assert_fail () from /usr/lib/libc.so.6
2018-09-24 21:42:26.237  Error: #9  0x0000558e8d7c4ddd in boost::signals2::mutex::lock (this=0x558e8f0cdb40) at /usr/include/boost/signals2/detail/lwm_pthreads.hpp:60
2018-09-24 21:42:26.237  Error: #10 0x0000558e8d7d1cdc in boost::signals2::detail::unique_lock<boost::signals2::mutex>::unique_lock (this=0x7fd3ab7fd928, m=...) at /usr/include/boost/signals2/detail/unique_lock.hpp:29
2018-09-24 21:42:26.237  Error: #11 0x0000558e8d7cd5ea in boost::signals2::detail::garbage_collecting_lock<boost::signals2::mutex>::garbage_collecting_lock (this=0x7fd3ab7fd870, m=...) at /usr/include/boost/signals2/connection.hpp:43
2018-09-24 21:42:26.237  Error: #12 0x0000558e8d9486c4 in boost::signals2::detail::signal_impl<void (CDomoticzHardwareBase*), boost::signals2::optional_last_value<void>, int, std::less<int>, boost::function<void (CDomoticzHardwareBase*)>, boost::function<void (boost::signals2::connection const&, CDomoticzHardwareBase*)>, boost::signals2::mutex>::operator()(CDomoticzHardwareBase*) (this=0x558e8f0cda00, args#0=0x558e8f073950) at /usr/include/boost/signals2/detail/signal_template.hpp:230
2018-09-24 21:42:26.237  Error: #13 0x0000558e8d948136 in boost::signals2::signal<void (CDomoticzHardwareBase*), boost::signals2::optional_last_value<void>, int, std::less<int>, boost::function<void (CDomoticzHardwareBase*)>, boost::function<void (boost::signals2::connection const&, CDomoticzHardwareBase*)>, boost::signals2::mutex>::operator()(CDomoticzHardwareBase*) (this=0x558e8f0739e0, args#0=0x558e8f073950) at /usr/include/boost/signals2/detail/signal_template.hpp:722
2018-09-24 21:42:26.237  Error: #14 0x0000558e8dbd9204 in CTellstick::StartHardware (this=0x558e8f073950) at /home/risca/Documents/programming/domoticz/hardware/Tellstick.cpp:221
2018-09-24 21:42:26.237  Error: #15 0x0000558e8d9ccbaf in CDomoticzHardwareBase::Start (this=0x558e8f073950) at /home/risca/Documents/programming/domoticz/hardware/DomoticzHardware.cpp:32
2018-09-24 21:42:26.237  Error: #16 0x0000558e8d789f71 in MainWorker::StartDomoticzHardware (this=0x558e8e5188a0 <m_mainworker>) at /home/risca/Documents/programming/domoticz/main/mainworker.cpp:298
2018-09-24 21:42:26.237  Error: #17 0x0000558e8d79147e in MainWorker::Do_Work (this=0x558e8e5188a0 <m_mainworker>) at /home/risca/Documents/programming/domoticz/main/mainworker.cpp:1579
2018-09-24 21:42:26.237  Error: #18 0x0000558e8d7eafe3 in std::__invoke_impl<void, void (MainWorker::*)(), MainWorker*> (__f=@0x558e8f13f330: (void (MainWorker::*)(MainWorker * const)) 0x558e8d7911b0 <MainWorker::Do_Work()>, __t=@0x558e8f13f328: 0x558e8e5188a0 <m_mainworker>) at /usr/include/c++/8.2.1/bits/invoke.h:73
2018-09-24 21:42:26.237  Error: #19 0x0000558e8d7e7ee4 in std::__invoke<void (MainWorker::*)(), MainWorker*> (__fn=@0x558e8f13f330: (void (MainWorker::*)(MainWorker * const)) 0x558e8d7911b0 <MainWorker::Do_Work()>, __args#0=@0x558e8f13f328: 0x558e8e5188a0 <m_mainworker>) at /usr/include/c++/8.2.1/bits/invoke.h:95
2018-09-24 21:42:26.237  Error: #20 0x0000558e8d7f09bf in std::thread::_Invoker<std::tuple<void (MainWorker::*)(), MainWorker*> >::_M_invoke<0ul, 1ul> (this=0x558e8f13f328) at /usr/include/c++/8.2.1/thread:244
2018-09-24 21:42:26.237  Error: #21 0x0000558e8d7f024d in std::thread::_Invoker<std::tuple<void (MainWorker::*)(), MainWorker*> >::operator() (this=0x558e8f13f328) at /usr/include/c++/8.2.1/thread:253
2018-09-24 21:42:26.237  Error: #22 0x0000558e8d7ef76a in std::thread::_State_impl<std::thread::_Invoker<std::tuple<void (MainWorker::*)(), MainWorker*> > >::_M_run (this=0x558e8f13f320) at /usr/include/c++/8.2.1/thread:196
2018-09-24 21:42:26.237  Error: #23 0x0000558e8de5a213 in execute_native_thread_routine ()
2018-09-24 21:42:26.237  Error: #24 0x00007fd3b6926a9d in start_thread () from /usr/lib/libpthread.so.0
2018-09-24 21:42:26.237  Error: #25 0x00007fd3b66b1a43 in clone () from /usr/lib/libc.so.6
Aborted (core dumped)
```

**Comment by xbeaudouin** [comment]
I really think we should keep the local detection of telldus libraries. 
Some OS have already the package and modification for the support of telldus has been done with the OS itself.
We can do as we already do with sqlite / mqtt the possibility to use "non builtin" telldus support for users that compile itself domoticz (in fact almost all telldus with domoticz, currently compile itself domoticz).
For example on FreeBSD telldus port there is *several* patches added to fix and adapt the software to the FreeBSD standard, and fix some issues: https://svnweb.freebsd.org/ports/head/comms/telldus-core/files/

I don't think we should maintain telldus support for all OS that domoticz support (more or less).

Regards

**Comment by Risca** [comment]
@xbeaudouin, I'm not sure if I fully understand you. The `WITH_TELLDUS` detection, at build time, still remains so if someone would compile domoticz themselves, it would link towards the library available at link time - just like before.
On systems that support `dlopen()`, and domoticz was compiled without `WITH_TELLDUS` enabled, it will try to detect telldus-core at runtime. Unless the FreeBSD port has made changes to the telldus API (didn't look like it), domoticz will use the system telldus-core (with fixes) on FreeBSD systems. If telldus-core symbols cannot be located, Tellstick devices won't work. That's how it is today: you can select Tellstick in the menu, but it won't work unless you build domoticz from source and link with telldus-core.

Basically, I just want to be able to use my Tellstick Duo on my raspberry pi, without having to recompile domoticz every time I want to update it. Since the domoticz build farm does not have, or want to have, telldus-core as a dependency (I can understand that), I opted for trying to get this approach merged upstream. The added benefit is that `Tellstick.h` and `Tellstick.cpp` will now always be compiled, and if someone breaks them it will be caught by the CI system.

**Comment by xbeaudouin** [comment]
I had read very fast the PR when I made the comment. In fact it should work... Well if the FreeBSD part have issues, I will make PR to fix that. I think this is ok for the FreeBSD part anyway.

**Comment by gizmocuz** [comment]
Got some compile error for windows, see the appveyor log:

```
mainworker.obj : error LNK2001: unresolved external symbol "public: static bool __cdecl CTellstick::Create(class CTellstick * *,int,int,int)" (?Create@CTellstick@@SA_NPAPAV1@HHH@Z) [C:\projects\domoticz\msbuild\domoticz.vcxproj]
C:\projects\domoticz\msbuild\Release\domoticz.exe : fatal error LNK1120: 1 unresolved externals
```

**Comment by Risca** [comment]
Not sure how to fix the Windows linking error. I added _hardware/TellstickFactory.cpp_ to the `domoticz_SRCS` set for CMake, but I don't see it being compiled in the appveyor log.
I could spin up a virtual machine and investigate later this week. What is the recommended Windows version?

**Comment by Risca** [comment]
Okay, the Windows build is using `domoticz.vcxproj` files from the msbuild folder. I did an attempt at updating them, but I might need to setup a real environment to get it right. We'll see what appveyor thinks

**Comment by Risca** [comment]
My latest 2 commits seems to make AppVeyor happy =)

Do I need to update anything else to get this merged?

**Comment by gizmocuz** [comment]
No this seems to compile fine.
But 180 new files in domoticz is a no go.

I would like to see a voting result of how many people are using the tellstick, because from what i understand is that this can be replaced with the RFLink and RFXTrx transceivers and those do not need such a big API/SDK inclusion in domoticz and support all sensor types (if not more)
The same happened with supporting the raspberry pi zwave addon, nice hardware !, but replaced by a 30$ stick which is supported by more software then domoticz. (and it eliminates a lot of duplicate code)

The impact of including the whole SDK (PR has a change of 180 files!) will make this repository large.

Maybe we could add an additional git project dependency and then add a compile directive to build support for tellstick... but that is basically what is already done now

Could this whole SDK be simplified ? Or can it be replaced by using a $30 RFLink ?

**Comment by Risca** [comment]
The SDK was already checked in - 180 changed files is me removing the SDK. It wasn't even used by the build system

**Comment by Risca** [comment]
Basically, what I did was:
* remove the checked in telldus core files that were only taking up space
* break out the dependency for telldus core only to TellstickFactory. If the factory can detect telldus core, Tellstick support will work as before

**Comment by gizmocuz** [comment]
Sorry i can see that now ;) thanks !!
